lldb/simulator_focus.py

```diff
@@ -0,0 +1,38 @@
+import subprocess
+import threading
+import lldb
+
+def focusSimulator(udid):
+    subprocess.Popen(["open", "-a", "Simulator", "--args", "-CurrentDeviceUDID", udid])
```

Comment:
I don't think you need `-CurrentDeviceUDID`, as I mentioned in the other comment it just sets the last simulator for when the simulator is launching the last one... You could add `-StartLastDeviceOnLaunch 0` just in case, although I think all those args do nothing when the app is already launched. This way you can just remove `udid` from this file altogether if you wanted.

One thing I noted, I happened to kill the simulator while debugging, which stops the debugger on throw by default, and then resuming the debugger launches the simulator back before the debug session finishes... not sure if it is easy to fix, but if there is a way to only call open if the pid being debugged is still listed as running... no big deal though.

Reply:
We'd need some way to specify which simulator to focus in case multiple simulators and running. Not sure if `-CurrentDeviceUDID` is doing this or not, need to try.

Reply:
I was actually just playing with that, Xcode does the right thing, but `open` opens the last one opened... `-CurrentDeviceUDID` just tells Simulator which one to boot if `StartLastDeviceOnLaunch` is enabled, does nothing if Simulator is opened already... I think the way to make this work is with `osascript`, which I think can focus a window by name with the right incantation (and won't open the app if already closed), but I have no experience with it at all... would like to know how Xcode focuses the right one.

Reply:
I tried a couple of things and managed to put together a script which focuses relevant simulator window, but haven't managed to run that script from the python instance in the lldb (maybe some permission issues?).

If we can't focus the correct simulator, I think we will just end up annoying the user. Worst case, we can enable the focus behaviour if only one simulator is open and disable it when multiple simulator are open. Thought it is not ideal and may just confuse users. For now, I'm planning to commit the part where we open the simulator without focus on boot and keep working on getting the correct behaviour for focusing simulator on attach and continue. Inputs are more than welcome!

Given the title, the following function should focus the relevant window. Posting here for reference.
```
def focusSimulator(title):
    cmd = """
tell application "System Events" to tell process "Simulator"
    windows where title contains "%s"
    if (count of result) is 1 then
        perform action "AXRaise" of item 1 of result
        do shell script "open -a Simulator --args -StartLastDeviceOnLaunch 0"
    end if
end tell
    """ % title

    p = subprocess.Popen(["osascript", "-"], stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)
    stdout, stderr = p.communicate(cmd)
```

Reply:
Thanks for looking into this! I had been playing with it too out of curiosity, Apple Script is horrible to work with lol. Your problem is that VSCode needs to have `Accessibility` permissions. If you see the LLDB output, it has the error `66:110: execution error: System Events got an error: osascript is not allowed assistive access. (-25211)` (sometimes it doesn't show it).

Another issue I noticed with Apple Script, the `process "Simulator"` command won't see windows across spaces (so if VSCode is full screen but Simulator is not, it won't find the window)... I think this is a pretty good compromise:

```
def focusSimulator(title):
    raise_simulator_window = """
tell application "System Events" to tell process "Simulator"
    windows where title contains "%s"
    if (count of result) is 1 then
        perform action "AXRaise" of item 1 of result
    end if
end tell
    """ % title
    subprocess.call(["osascript", "-e", raise_simulator_window])
    subprocess.call(["osascript", "-e", "activate application \"Simulator\""])
```

Few notes about the code above:
1. Only the first command needs `Accessibility` permissions for VSCode, so if a user doesn't have them set at least the Simulator will focus (and for most cases with only one open, or the relevant one opened last, it will do the right thing).
2. If VSCode and Simulator are in different spaces, it will still focus the Simulator (it will just not raise the right window first, as the first command won't find the window), so it is a pretty good fallback for most cases like in point 1.
3. It doesn't reopen the Simulator if it was closed/killed during debugging, since `activate` only focuses if it is opened.

I know it is still not great, so do with this what you like :) I don't know if there are better tools for this, pretty surprising how hard it is to do.

Reply:
Thanks for this, @bzarco!

I did some more playaround and came up with a slightly new approach to the AppleScript. It is now essentially trying to emulate pressing the Simulator -> Window -> <simulator name> button in the menu to activate the correct simulator. This seems to work pretty well for me in most cases. Of course some annoyances are left and I'm not sure how fragile this is. Have a look if you want, it'll help to have a second set of eyes.

Also, seems like Xcode also doesn't handle the multiple simulator case pretty well. On launch, it focuses the correct simulator (although, it fails to do so in some cases), and on continue, it always focuses the last focused simulator, regardless of the currently debugged one (essentially, `activate application "Simulator"`).